packages/plugin-data/src/commands/force/data/soql/query.ts

```diff
@@ -0,0 +1,198 @@
+            const f: Field = {
+              fieldType: FieldType.field,
+              name: ensureString(ensureJsonMap(subcolumn).columnName),
+            };
+            if (field.fields) field.fields.push(f);
+          }
+          columns.push(field);
+        } else {
+          for (const subcolumn of column.joinColumns) {
+            const f: Field = {
+              fieldType: FieldType.field,
+              name: `${name}.${ensureString(ensureJsonMap(subcolumn).columnName)}`,
+            };
+            columns.push(f);
+          }
+        }
+      } else if (column.aggregate) {
+        const field: Field = {
+          fieldType: FieldType.functionField,
+          name: ensureString(column.displayName),
+        };
+        // If it isn't an alias, skip so the display name is used when messaging rows
+        if (!/expr[0-9]+/.test(name)) {
+          field.alias = name;
+        }
+        columns.push(field);
+      } else {
+        columns.push({ fieldType: FieldType.field, name } as Field);
+      }
+    }
+    return columns;
+  }
+}
+
+export class DataSoqlQueryCommand extends SfdxCommand {
+  public static readonly description = messages.getMessage('description');
+  public static readonly requiresProject = false;
+  public static readonly requiresUsername = true;
+  public static readonly examples = messages.getMessage('examples').split(os.EOL);
+
+  public static readonly flagsConfig: FlagsConfig = {
+    query: flags.string({
+      char: 'q',
+      required: true,
+      hidden: false,
+      description: messages.getMessage('queryToExecute'),
+      longDescription: messages.getMessage('queryLongDescription'),
+    }),
+    usetoolingapi: flags.boolean({
+      char: 't',
+      required: false,
+      hidden: false,
+      default: false,
+      description: messages.getMessage('queryToolingDescription'),
+    }),
+    resultformat: flags.enum({
+      char: 'r',
+      description: messages.getMessage('resultFormatDescription'),
+      longDescription: messages.getMessage('resultFormatLongDescription'),
+      options: ['human', 'csv', 'json'],
+      default: 'human',
+    }),
+  };
+
+  // Overrides SfdxCommand.  This is ensured since requiresUsername == true
+  protected org!: Org;
+
+  /**
+   * Command run implementation
+   *
+   * Returns either a DataSoqlQueryResult or a SfdxResult.
+   * When the user is using global '--json' flag an instance of SfdxResult is returned.
+   * This is necessary since '--json' flag reports results in the form of SfdxResult
+   * and bypasses the definition of start result. The goal is to have the output
+   * from '--json' and '--resulformat json' be the same.
+   *
+   * The DataSoqlQueryResult is necessary to communicate user selections to the reporters.
+   * The 'this' object available during display() function does not include user input to
+   * the command, which are necessary for reporter selection.
+   *
+   */
+  public async run(): Promise<SfdxResult> {
```

Comment:
We do not want to return `SfdxResults`. If you look at the existing `--json` output, there is not a `data` property. It should be `QueryResult<Record>` from jsforce.